Add org.freedesktop.Sdk.Extension.swift5

david-swift opened the pull request:
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I have [built][build] and tested the submission locally.
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [ ] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub.
- [ ] The domain used for the application ID is controlled by the application developers either directly or through the code hosting (e.g. GitHub, GitLab, SourceForge, etc.). The [application id guidelines][app-id] are followed.
- [ ] Any additional patches or files have been submitted to the upstream projects concerned.

Do the last three todos also apply to SDK extensions?

Thanks for reviewing already in advance!

[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[maint]: https://docs.flathub.org/docs/for-app-authors/maintanance
[submission]: https://docs.flathub.org/docs/for-app-authors/submission
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission
[app-id]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id

hfiguiere commented on org.freedesktop.Sdk.Extension.swift.json:
```diff
@@ -0,0 +1,206 @@
+            "sources": [
+                {
+                    "type": "archive",
+                    "url": "https://download.swift.org/swift-5.9.1-release/centos7/swift-5.9.1-RELEASE/swift-5.9.1-RELEASE-centos7.tar.gz",
```
how will this work on aarch64?

hfiguiere replied:
also you can take something more recent that centos7. The Freedesktop SDK is a pretty recent stack

david-swift replied:
I switched to Ubuntu 22.04, which has an x86_64 and an aarch64 download.

david-swift replied:
I switched once again, this time to the Amazon Linux 2 downloads, because I could remove the "really really dirty" part and the lines that fail on aarch64 that way.

hfiguiere commented on org.freedesktop.Sdk.Extension.swift.json:
```diff
@@ -0,0 +1,206 @@
+            "build-commands": [
+                "mkdir -p /usr/lib/sdk/swift",
+                "cp -R usr/* /usr/lib/sdk/swift",
+                "sed -i s/libedit.so.2/libedit.so.0/g /usr/lib/sdk/swift/bin/lldb-server",
```
this is really really dirty to use sed to change the soversion because they don't match.

hfiguiere commented on org.freedesktop.Sdk.Extension.swift.json:
```diff
@@ -0,0 +1,213 @@
+                "sed -i s/libedit.so.2/libedit.so.0/g /usr/lib/sdk/swift/lib/liblldb.so",
+                "sed -i s/libedit.so.2/libedit.so.0/g /usr/lib/sdk/swift/lib/liblldb.so.13git",
+                "ln -s /usr/lib/x86_64-linux-gnu/libtinfo.so.6 /usr/lib/sdk/swift/lib/libtinfo.so.5",
+                "ln -s /usr/lib/x86_64-linux-gnu/libncurses.so.6 /usr/lib/sdk/swift/lib/libncurses.so.5"
```
this will fail on aarch64

bbhtt commented:
What's the usecase for this on Flathub?

david-swift commented:
> What's the usecase for this on Flathub?

I'm working on [adwaita-swift](https://github.com/AparokshaUI/adwaita-swift) for creating GNOME apps using Swift in a readable, declarative way. I am e.g. working on [Flashcards](https://github.com/david-swift/Flashcards). The API is very similar to SwiftUI used for iOS/macOS development, so hopefully, it might attract some developers currently focused on Apple platforms into the Linux world. Swift could also be an option for other projects, as it offers good C and C++ interoperability.

hfiguiere commented:
any tooling to generate manifest for swift packages like we have of npn, cargo, pip?

hfiguiere commented:
> What's the usecase for this on Flathub?

There is at least one package on flathub that is built with Swift.

david-swift commented:
> any tooling to generate manifest for swift packages like we have of npn, cargo, pip?

There's the [Swift Package Manager](https://www.swift.org/documentation/package-manager/) which comes with the Swift download itself, so it is already included in this extension.

david-swift commented:
`swift package init` generates a starting point for a package including the manifest.

hfiguiere commented:
> There's the [Swift Package Manager](https://www.swift.org/documentation/package-manager/) which comes with the Swift download itself, so it is already included in this extension.

According to the doc:

> The role of the package manager is to reduce coordination costs by automating the process of downloading and building all of the dependencies for a project.

Flathub builds are done *offline* so none of this can happen during the build. 

"starting point" isn't really what I have in mind. Generating a manifest ready to use is what the tooling does (more or less with bugs and other).

hfiguiere commented:
bot, build org.freedesktop.Sdk.Extension.swift

hfiguiere commented on org.freedesktop.Sdk.Extension.swift.json:
```diff
@@ -0,0 +1,191 @@
+    "build-options": {
+        "prefix": "/usr/lib/sdk/swift",
+        "prepend-pkg-config-path": "/usr/lib/sdk/swift/lib/pkgconfig/",
+        "prepend-path": "/usr/lib/sdk/swift/bin/:/usr/lib/sdk/llvm14/bin",
```
extension is llvm16

hfiguiere commented:
what's the plan for future version of the toolchain? It has been my understanding that Swift version brought need to update code. llvm use major versions as part of the package name. So does JDK. Or dotNET.

hfiguiere commented:
Also it would be nice it have an appstream file so the version can be displayed where appropriate.

david-swift commented:
> > There's the [Swift Package Manager](https://www.swift.org/documentation/package-manager/) which comes with the Swift download itself, so it is already included in this extension.
> 
> According to the doc:
> 
> > The role of the package manager is to reduce coordination costs by automating the process of downloading and building all of the dependencies for a project.
> 
> Flathub builds are done _offline_ so none of this can happen during the build.
> 
> "starting point" isn't really what I have in mind. Generating a manifest ready to use is what the tooling does (more or less with bugs and other).

Ah, you're talking about a Flatpak manifest file? In this extension, there are also two Swift packages built, and its dependencies were added manually to the manifest file. I think it makes sense to automate this, so I'll look into implementing this.

hfiguiere commented:
Not only for this extension. But for packagers using it. Usually these "always online" package managers are the biggest barrier. While our tooling isn't perfect https://github.com/flatpak/flatpak-builder-tools , it sure beat no support.

hfiguiere commented on org.freedesktop.Sdk.Extension.swift.json:
```diff
@@ -0,0 +1,191 @@
+    "build-options": {
+        "prefix": "/usr/lib/sdk/swift",
+        "prepend-pkg-config-path": "/usr/lib/sdk/swift/lib/pkgconfig/",
+        "prepend-path": "/usr/lib/sdk/swift/bin/:/usr/lib/sdk/llvm16/bin",
```
funny thing the build seems to have passed even with the wrong path.

david-swift commented:
Just to be sure the message gets to you, @hfiguiere (you were _incredibly_ fast with the other reviews, so I'm not quite sure whether you saw the review request I opened yesterday - don't feel stressed though 😀): I renamed the extension to `org.swift.freedesktop.Sdk.Extension.swift5` and removed the useless build options so I think the PR would be ready for another review/test build.

david-swift commented:
bot, build org.freedesktop.Sdk.Extension.swift5

david-swift commented:
@hfiguiere Just realized I can request a build as well. I tested this build, it is working as expected.

bbhtt commented on org.freedesktop.Sdk.Extension.swift5.metainfo.xml:
```diff
@@ -0,0 +1,23 @@
+  <metadata_license>MIT</metadata_license>
+  <name>Swift 5</name>
+  <summary>Swift 5 compiler and tools for the Flatpak Freedesktop SDK</summary>
+  <description>This extension adds Swift support to Flatpak.</description>
```
```suggestion
  <description>This extension adds Swift support to Flatpak</description>
```

david-swift commented:
Thank you!
